File: Services/ChannelSlotPolicy.swift
import Foundation

/// 「保存できるチャンネルは何件か」だけを決める純粋なロジック。
///
/// 制限するのは**同時に保存できるチャンネル数だけ**で、保存したチャンネルの中の機能は
/// 無料でも一切制限しない（並べ替え・視聴済み・続きから・移動ボタン・メモ・進捗・公式プレイヤー再生）。
///
/// Android 版 `billing/ChannelSlotPolicy.kt` と同じ判定にしてある（挙動を揃えるため）。
enum ChannelSlotPolicy {

    /// 無料で同時に保存できるチャンネル数。
    static let freeChannelLimit = 1

    /// そのチャンネルを新しく開いて（＝保存して）よいか。
    ///
    /// - Pro なら常に可
    /// - すでに保存済みなら可（枠を増やさないため）
    /// - 空きがあれば可
    static func canOpen(savedChannelIds: [String], channelId: String, isPro: Bool) -> Bool {
        isPro || savedChannelIds.contains(channelId) || savedChannelIds.count < freeChannelLimit
    }

    /// いま使えるチャンネル。ここに入らない保存済みチャンネルは**ロック**する。
    ///
    /// Pro を返金・取消などで失った人が、保存済みの複数チャンネルをそのまま使い続けられると
    /// 「買って返して使い放題」になるため、上限を超えた分は開けなくする。
    /// ただし**記録は消さない**。Pro を買い直せば全部そのまま戻る。
    ///
    /// - Parameters:
    ///   - savedChannelIdsNewestFirst: 最終利用が新しい順の保存済みチャンネル
    ///   - activeChannelId: 利用者が「無料ではこれを使う」と選んだチャンネル（未選択なら nil）
    static func usableChannelIds(savedChannelIdsNewestFirst: [String],
                                 isPro: Bool,
                                 activeChannelId: String?) -> Set<String> {
        if isPro || savedChannelIdsNewestFirst.count <= freeChannelLimit {
            return Set(savedChannelIdsNewestFirst)
        }
        var usable: [String] = []
        if let activeChannelId, savedChannelIdsNewestFirst.contains(activeChannelId) {
            usable.append(activeChannelId)
        }
        // 未選択・枠が余っている場合は、最後に開いたのが新しい方から埋める。
        for id in savedChannelIdsNewestFirst where usable.count < freeChannelLimit {
            if !usable.contains(id) { usable.append(id) }
        }
        return Set(usable)
    }

    /// 保存はされているが、いまは開けない状態か。
    static func isLocked(savedChannelIdsNewestFirst: [String],
                         channelId: String,
                         isPro: Bool,
                         activeChannelId: String?) -> Bool {
        guard savedChannelIdsNewestFirst.contains(channelId) else { return false }
        return !usableChannelIds(savedChannelIdsNewestFirst: savedChannelIdsNewestFirst,
                                 isPro: isPro,
                                 activeChannelId: activeChannelId).contains(channelId)
    }

    /// 無料ユーザーが「入れ替える」を選んだときに外すチャンネル。
    /// 新しく1件入れる空きを作る分だけ、最後に開いたのが古い方から外す。
    ///
    /// - Important: 外したチャンネルの記録は**削除する**（`ChannelDataRemover`）。
    ///   ロック（Pro 失効時）とは扱いが違うので混同しないこと。
    static func idsToRemoveForReplacement(savedChannelIdsNewestFirst: [String]) -> [String] {
        let keep = max(freeChannelLimit - 1, 0)
        guard savedChannelIdsNewestFirst.count > keep else { return [] }
        return Array(savedChannelIdsNewestFirst.dropFirst(keep))
    }
}

File: android/app/src/main/java/com/deskflowlabs/channeltimelineviewer/billing/ChannelSlotPolicy.kt
package com.deskflowlabs.channeltimelineviewer.billing

/**
 * 「保存できるチャンネルは何件か」だけを決める純粋なロジック。
 *
 * 制限するのは**同時に保存できるチャンネル数だけ**で、保存済みチャンネルの中の機能は
 * 無料でも一切制限しない（並べ替え・視聴済み・続きから・メモ・進捗・公式プレイヤー再生）。
 *
 * Google Play や Compose に依存させないので、そのまま単体テストできる。
 */
object ChannelSlotPolicy {

    /** 無料で同時に保存できるチャンネル数。 */
    const val FREE_CHANNEL_LIMIT = 1

    /**
     * そのチャンネルを開いて（＝保存して）よいか。
     *
     * - Pro なら常に可
     * - すでに保存済みのチャンネルなら可（枠を増やさないため）
     * - 空きがあれば可
     *
     * 以前のバージョンで2件以上保存している人（この課金を入れる前からの利用者）が
     * いても、**保存済みのものは開ける**。減らすのは本人の操作に任せる。
     */
    fun canOpen(savedChannelIds: Collection<String>, channelId: String, isPro: Boolean): Boolean =
        isPro || channelId in savedChannelIds || savedChannelIds.size < FREE_CHANNEL_LIMIT

    /**
     * いま使えるチャンネル（それ以外はロックする）。
     *
     * Pro を返金・解約などで失った人が、保存済みの複数チャンネルをそのまま使い続けられると
     * 「買って返して使い放題」になってしまうので、上限を超えた分は**開けなくする**。
     * ただし**記録は消さない**。Pro を買い直せば全部そのまま戻る。
     *
     * @param savedChannelIdsNewestFirst 最終利用が新しい順に並んだ保存済みチャンネル
     * @param activeChannelId 利用者が「これを使う」と選んだチャンネル（未選択なら null）
     */
    fun usableChannelIds(
        savedChannelIdsNewestFirst: List<String>,
        isPro: Boolean,
        activeChannelId: String?,
    ): Set<String> {
        if (isPro || savedChannelIdsNewestFirst.size <= FREE_CHANNEL_LIMIT) {
            return savedChannelIdsNewestFirst.toSet()
        }
        val usable = LinkedHashSet<String>()
        activeChannelId?.takeIf { it in savedChannelIdsNewestFirst }?.let(usable::add)
        // 選ばれていない・足りない分は、最後に開いたのが新しい方から埋める。
        for (id in savedChannelIdsNewestFirst) {
            if (usable.size >= FREE_CHANNEL_LIMIT) break
            usable.add(id)
        }
        return usable
    }

    /** そのチャンネルがロックされているか（保存はされているが、いまは開けない）。 */
    fun isLocked(
        savedChannelIdsNewestFirst: List<String>,
        channelId: String,
        isPro: Boolean,
        activeChannelId: String?,
    ): Boolean = channelId in savedChannelIdsNewestFirst &&
        channelId !in usableChannelIds(savedChannelIdsNewestFirst, isPro, activeChannelId)

    /**
     * 入れ替えのために外すチャンネル。新しく1件入れる空きを作る分だけ、
     * 古い（最後に開いたのが古い）ものから外す。
     *
     * @param savedChannelIdsNewestFirst 最終利用が新しい順に並んだ保存済みチャンネル
     */
    fun idsToRemoveForReplacement(savedChannelIdsNewestFirst: List<String>): List<String> {
        val keep = (FREE_CHANNEL_LIMIT - 1).coerceAtLeast(0)
        if (savedChannelIdsNewestFirst.size <= keep) return emptyList()
        return savedChannelIdsNewestFirst.drop(keep)
    }
}
